### PR TITLE
Implemented VirtualMediaActions for Dell plugin

### PR DESCRIPTION
**Pull request by amar-shalgar**
Added support for VirtualMediaActions in Dell plugin
Updated UT
After VirtualMediaAction is success saving the latest response in DB

### CODE REVIEW

**Review by jeevan-kamkar — APPROVED** [review]

File: plugin-dell/dphandler/managers.go [hunk 1]
```diff
@@ -168,3 +170,91 @@ func getInfoFromDevice(uri string, deviceDetails dpmodel.Device, ctx iris.Contex
...
+		ctx.StatusCode(statusCode)
+		ctx.WriteString(errMsg)
+		return
+	}
+
+	if statusCode == http.StatusNoContent {
+		log.Info("VirtualMediaActions is successful for URI ")
+		statusCode = http.StatusOK
+		body = vmActionsResponse()
+	}
```
Comment (bharath-b-hpe) [inline plugin-dell/dphandler/managers.go:229]: Please handle error conditions.

File: svc-managers/managers/managers.go [hunk 2]
```diff
@@ -154,7 +154,7 @@ func (e *ExternalInterface) GetManagers(req *managersproto.ManagerRequest) respo
 					return resp
 				}
 			}
-			err = e.DB.UpdateManagersData(req.URL, managerData)
+			err = e.DB.UpdateManagersData(req.URL, managerData, "Managers")
```
Comment (bharath-b-hpe) [inline svc-managers/managers/managers.go]: We are updating managers data, do we need to pass table name, isn't implicit?

File: plugin-dell/dphandler/managers.go [hunk 3]
```diff
@@ -168,3 +170,91 @@ func getInfoFromDevice(uri string, deviceDetails dpmodel.Device, ctx iris.Contex
...
+	if err != nil {
+		errMsg := "while performing actions on virtual media, got: " + err.Error()
+		log.Error(errMsg)
+		ctx.StatusCode(statusCode)
+		ctx.WriteString(errMsg)
+		return
+	}
+
+	if statusCode == http.StatusNoContent {
+		log.Info("VirtualMediaActions is successful for URI ")
```
Comment (Bharath-KKB) [inline plugin-dell/dphandler/managers.go]: what action failed and for what URI ? Log those as well. Dont we have specific actions - VirtualMedia.InsertMedia and VirtualMedia.EjectMedia ?

File: svc-managers/managers/managers.go [hunk 4]
```diff
@@ -313,10 +313,29 @@ func (e *ExternalInterface) VirtualMediaActions(req *managersproto.ManagerReques
 	}
 	// splitting managerID to get uuid
 	requestData := strings.Split(req.ManagerID, ":")
-	log.Info(requestData)
 	uuid := requestData[0]
-	log.Info(req.URL, uuid, string(requestBody))
 	resp = e.deviceCommunication(req.URL, uuid, requestData[1], http.MethodPost, requestBody)
+
+	// If the virtualmedia action is success then updating DB
+	if resp.StatusCode == http.StatusOK {
```
Comment (bharath-b-hpe) [inline svc-managers/managers/managers.go:320]: Please handle error scenarios too.

File: plugin-dell/main.go [hunk 5]
```diff
@@ -201,6 +201,8 @@ func routers() *iris.Application {
 		managers.Get("/{id}/HostInterfaces/{rid}", dphandler.GetResource)
 		managers.Get("/{id}/VirtualMedia", dphandler.GetResource)
 		managers.Get("/{id}/VirtualMedia/{rid}", dphandler.GetResource)
+		managers.Post("/{id}/VirtualMedia/{rid}/Actions/VirtualMedia.EjectMedia", dphandler.VirtualMediaActions)
```
Comment (irappa-pattar) [inline plugin-dell/main.go:204]: please run go tools

**Review by irappa-pattar — CHANGES_REQUESTED** [review]

File: plugin-dell/dphandler/managers.go [hunk 6]
```diff
@@ -168,3 +170,91 @@ func getInfoFromDevice(uri string, deviceDetails dpmodel.Device, ctx iris.Contex
...
+		statusCode = http.StatusOK
+		body = vmActionsResponse()
+	}
+
+	ctx.StatusCode(statusCode)
+	ctx.Write(body)
+}
+
+// vmActionsResponse response is used for creating a final response for virtual media actions
+func vmActionsResponse() []byte {
```
Comment (Bharath-KKB) [inline plugin-dell/dphandler/managers.go]: This is an ambiguous name rename to createVirtMediaActionResponse

File: svc-managers/mgrmodel/manager.go [hunk 7]
```diff
@@ -161,7 +161,7 @@ func GetManagerByURL(url string) (string, *errors.Error) {
 }
 
 // UpdateManagersData will modify the current details to given changes
-func UpdateManagersData(key string, managerData map[string]interface{}) error {
+func UpdateManagersData(key string, managerData map[string]interface{}, table string) error {
```
Comment (bharath-b-hpe) [inline svc-managers/mgrmodel/manager.go]: We don't need to pass the table name, as the method is for updating Managers data.

File: svc-managers/mgrmodel/manager_test.go [hunk 8]
```diff
@@ -61,7 +61,7 @@ func TestManager_Update(t *testing.T) {
 			"State": "Absent",
 		},
 	}
-	err = UpdateManagersData(key, m)
+	err = UpdateManagersData(key, m, "Managers")
 
 	data, err := GetResource(table, key)
```
Comment (Bharath-KKB) [inline svc-managers/mgrmodel/manager_test.go:66]: This method GetResource only returns "body" if key is not /redfish/v1/Managers/uuid1:1/Ethernet. So what are we testing here ??

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
Address comments

File: svc-managers/managers/managers.go (hunk 4 above)
Comment (amar-shalgar) [reply svc-managers/managers/managers.go:320]: This is only for updating DB, there is nothing to update in DB if the operation is failed

File: svc-managers/managers/managers.go (hunk 2 above)
Comment (amar-shalgar) [reply svc-managers/managers/managers.go]: Same function is used for other table updates under manager api

File: svc-managers/mgrmodel/manager.go (hunk 7 above)
Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager.go]: Same function is used for other table updates under manager api

File: svc-managers/mgrmodel/manager_test.go (hunk 8 above)
Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager_test.go:66]: GetResource is getting the data from the key and later in the code cross checking if the data got updated or not

Comment (Bharath-KKB) [reply svc-managers/mgrmodel/manager_test.go:66]: I meant mockGetResource  which is called when we call GetResource - check the definition and see if I am making sense

File: svc-managers/mgrmodel/manager.go [hunk 9]
```diff
@@ -160,19 +160,19 @@ func GetManagerByURL(url string) (string, *errors.Error) {
 	return manager, nil
 }
 
-// UpdateManagersData will modify the current details to given changes
-func UpdateManagersData(key string, managerData map[string]interface{}) error {
+// UpdateData will modify the current details to given changes
+func UpdateData(key string, managerData map[string]interface{}, table string) error {
```
Comment (Bharath-KKB) [inline svc-managers/mgrmodel/manager.go]: Why are we changing these functions? Was there an architecture discussion? BB why isnt team catching these in review ?

Comment (Bharath-KKB) [reply svc-managers/mgrmodel/manager.go]: managerData looks familiar ?

File: plugin-dell/dphandler/managers_test.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+	"github.com/ODIM-Project/ODIM/plugin-dell/config"
+	"github.com/ODIM-Project/ODIM/plugin-dell/dpmodel"
+	"github.com/ODIM-Project/ODIM/plugin-dell/dpresponse"
+	iris "github.com/kataras/iris/v12"
+	"github.com/kataras/iris/v12/httptest"
+)
+
+func mockManagers(username, password, url string, w http.ResponseWriter) {
+	body := `{"data": "success"}`
+	if url == "/ODIM/v1/Managers/1/VirtualMedia/1/Actions/VirtualMedia.InsertMedia" && username == "admin" {
```
Comment (Bharath-KKB) [inline plugin-dell/dphandler/managers_test.go]: Do not hardcode /ODIM/v1 - use a const

File: svc-managers/managers/common.go [hunk 11]
```diff
@@ -43,7 +43,7 @@ type DB struct {
 	GetAllKeysFromTable func(string) ([]string, error)
 	GetManagerByURL     func(string) (string, *errors.Error)
 	GetPluginData       func(string) (mgrmodel.Plugin, *errors.Error)
-	UpdateManagersData  func(string, map[string]interface{}) error
+	UpdateData          func(string, map[string]interface{}, string) error
```
Comment (Bharath-KKB) [inline svc-managers/managers/common.go:46]: we should not change these functions so casually.

File: plugin-dell/main.go (hunk 5 above)
Comment (amar-shalgar) [reply plugin-dell/main.go:204]: Please check now

File: plugin-dell/dphandler/managers.go (hunk 1 above)
Comment (amar-shalgar) [reply plugin-dell/dphandler/managers.go:229]: Addressed the comments

File: svc-managers/managers/common_test.go [hunk 12]
```diff
@@ -132,7 +132,7 @@ func mockGetPluginData(pluginID string) (mgrmodel.Plugin, *errors.Error) {
 	}, nil
 }
 
-func mockUpdateManagersData(key string, managerData map[string]interface{}) error {
+func mockUpdateData(key string, managerData map[string]interface{}, table string) error {
```
Comment (Bharath-KKB) [inline svc-managers/managers/common_test.go]: How do we verify this action later ?

File: svc-managers/managers/managers.go (hunk 2 above)
Comment (amar-shalgar) [reply svc-managers/managers/managers.go]: Addressed comments

File: plugin-dell/dphandler/managers.go (hunk 6 above)
Comment (amar-shalgar) [reply plugin-dell/dphandler/managers.go]: Addressed comments

File: svc-managers/mgrmodel/manager.go (hunk 7 above)
Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager.go]: Addressed comments

File: svc-managers/managers/managers.go [hunk 13]
```diff
@@ -154,7 +154,7 @@ func (e *ExternalInterface) GetManagers(req *managersproto.ManagerRequest) respo
 					return resp
 				}
 			}
-			err = e.DB.UpdateManagersData(req.URL, managerData)
+			err = e.DB.UpdateData(req.URL, managerData, "Managers")
```
Comment (Bharath-KKB) [inline svc-managers/managers/managers.go:157]: What other data can we update from Managers module ?

File: svc-managers/mgrmodel/manager.go (hunk 9 above)
Comment (Bharath-KKB) [reply svc-managers/mgrmodel/manager.go]: As above what other data are we updating in manager module?

Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager.go]: Earlier this function was using a hardcoded value for update, I have reused the function by passing a parameter

Comment (Bharath-KKB) [reply svc-managers/mgrmodel/manager.go]: can i send a non managerData to this func ?

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
Address comments

File: svc-managers/managers/managers.go (hunk 13 above)
Comment (amar-shalgar) [reply svc-managers/managers/managers.go:157]: Now same function is used for updating virtualmedia data to DB

File: plugin-dell/dphandler/managers.go (hunk 3 above)
Comment (amar-shalgar) [reply plugin-dell/dphandler/managers.go]: Addressed the comments

Comment (amar-shalgar) [reply plugin-dell/dphandler/managers.go]: both actions are handled in the same function

File: svc-managers/mgrmodel/manager.go (hunk 9 above)
Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager.go]: Now this function can update data based on the key passed, earlier it was updating only manager key data

File: plugin-dell/dphandler/managers.go [hunk 14]
```diff
@@ -168,3 +170,101 @@ func getInfoFromDevice(uri string, deviceDetails dpmodel.Device, ctx iris.Contex
...
+		err = json.Unmarshal(device.PostBody, &reqBody)
+		if err != nil {
+			errMsg := "While unmarshaling the request to insert a virtual media, got: " + err.Error()
+			log.Error(errMsg)
+			ctx.StatusCode(http.StatusInternalServerError)
+			ctx.WriteString(errMsg)
+			return
+		}
+
+		device.PostBody, err = json.Marshal(reqBody)
```
Comment (irappa-pattar) [inline plugin-dell/dphandler/managers.go]: if you are not changing reqBody, why unmarshalling and marshalling again.

**Review by irappa-pattar — CHANGES_REQUESTED** [review]

File: svc-managers/mgrmodel/manager.go (hunk 7 above)
Comment (Bharath-KKB) [reply svc-managers/mgrmodel/manager.go]: But still the param name is managerData.

File: svc-managers/mgrmodel/manager.go (hunk 9 above)
Comment (Bharath-KKB) [reply svc-managers/mgrmodel/manager.go]: Then lets update paramname "managerData"

File: plugin-dell/dphandler/managers.go (hunk 14 above)
Comment (amar-shalgar) [reply plugin-dell/dphandler/managers.go]: This is done to make sure correct parameters come to plugin code

**Review by irappa-pattar — APPROVED** [review]

File: svc-managers/mgrmodel/manager.go (hunk 7 above)
Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager.go]: It is changed, here it is showing the outdated info, kindly check in the file

**Review by bharath-b-hpe — APPROVED** [review]

File: svc-managers/mgrmodel/manager.go (hunk 7 above)
Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager.go]: This looks outdated, please check in the file

File: svc-managers/mgrmodel/manager.go (hunk 9 above)
Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager.go]: Already the param name is changed, kindly check

File: plugin-dell/dphandler/managers_test.go (hunk 10 above)
Comment (amar-shalgar) [reply plugin-dell/dphandler/managers_test.go]: Addressed the comment

File: plugin-dell/dphandler/managers.go (hunk 3 above)
Comment (Bharath-KKB) [reply plugin-dell/dphandler/managers.go]: I dont understand your response. The comment was to log the action that failed.

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
fix mockUpdateData and mockGetResource implementation

File: svc-managers/rpc/manager_test.go [hunk 15]
```diff
@@ -108,7 +108,7 @@ func mockGetPluginData(pluginID string) (mgrmodel.Plugin, *errors.Error) {
 	}, nil
 }
 
-func mockUpdateManagersData(key string, managerData map[string]interface{}) error {
+func mockUpdateData(key string, updateData map[string]interface{}, table string) error {
```
Comment (Bharath-KKB) [inline svc-managers/rpc/manager_test.go:111]: This needs an implementation

File: svc-managers/mgrmodel/manager_test.go [hunk 16]
```diff
@@ -61,7 +61,7 @@ func TestManager_Update(t *testing.T) {
 			"State": "Absent",
 		},
 	}
-	err = UpdateManagersData(key, m)
+	err = UpdateData(key, m, "Managers")
 
 	data, err := GetResource(table, key)
```
Comment (Bharath-KKB) [inline svc-managers/mgrmodel/manager_test.go:66]: mockGetResourcxe also needs an implementation

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
Address remaining comments

File: plugin-dell/dphandler/managers.go (hunk 14 above)
Comment (Bharath-KKB) [reply plugin-dell/dphandler/managers.go]: As the params are sent elsewhere they will be verified there isnt it ? Even so why marshal this again ? We end up with these two operations on every request without a need.

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
Address comments

File: svc-managers/managers/common_test.go (hunk 12 above)
Comment (amar-shalgar) [reply svc-managers/managers/common_test.go]: Updated the mockUpdateData

File: plugin-dell/dphandler/managers.go (hunk 14 above)
Comment (amar-shalgar) [reply plugin-dell/dphandler/managers.go]: Addressed the comment

File: plugin-dell/dphandler/managers.go (hunk 3 above)
Comment (amar-shalgar) [reply plugin-dell/dphandler/managers.go]: Comments are incorporated, log is added for failure case as well

File: svc-managers/rpc/manager_test.go (hunk 15 above)
Comment (amar-shalgar) [reply svc-managers/rpc/manager_test.go:111]: Updated the mockUpdateData

File: svc-managers/mgrmodel/manager_test.go (hunk 8 above)
Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager_test.go:66]: Updated the mockGetResource

File: svc-managers/mgrmodel/manager_test.go (hunk 16 above)
Comment (amar-shalgar) [reply svc-managers/mgrmodel/manager_test.go:66]: Updated the mockGETResource

**Review by Bharath-KKB — APPROVED** [review]

**Pull request merged**